### PR TITLE
feat: add ReverseDNS automation module

### PR DESCRIPTION
**Pull request by KhadimBamba27**
## Description
This Pull Request adds the `ReverseDNS` automation module to the Sekoia library. This module allows the resolution of IP addresses (both public and private) into their corresponding hostnames using reverse DNS lookups (PTR requests).

## Context
This contribution is part of our official homologation request to enable this custom integration on our on-premise runners. 

Following the email exchange with Mr. DIENG, we have integrated our module files inside the `modules/reversedns/` directory.

## Components Added
- Global `manifest.json`
- Action configuration `action_reversedns_lookup.json`
- Python logic `action_reverse_dns.py` and entrypoint `main.py`
- Documentation and branding (`CHANGELOG.md`, `logo.png`)

Thank you in advance for your review and feedback!

## Summary by Sourcery

Add a new ReverseDNS automation module for performing reverse DNS lookups on IP addresses.

New Features:
- Introduce a ReverseDNS module exposing a reverse DNS lookup action to resolve IP addresses to hostnames.

Build:
- Add a dedicated Dockerfile and Poetry configuration for building and running the ReverseDNS module.

Documentation:
- Add changelog documentation describing the initial ReverseDNS module releases.

### CODE REVIEW

**Review by lvoloshyn-sekoia — CHANGES_REQUESTED** [review]
Hey @KhadimBamba27, I see some issues to fix:

1. File structure is incorrect: you don't need `modules/reversedns`, you need just `ReverseDNS` which should lie in the `automation_modules` directory. Look at [other modules](https://github.com/SEKOIA-IO/automation-library/pull/2683/changes) for examples.
2. Talking about the ReverseDNS name itself - I believe it makes sense to rename it to just DNS, as reverse DNS is only one of possible DNS actions. In the future, we could add more DNS-related stuff. Also I suggest to use file names like `action_dns_reverse_search` and class names like `DnsReverseSearchAction` for consistency.
3. You don't need a second `action_reversedns_lookup.json` in `reversedns_module`, only the one in the module's root.
4. `results` field in `action_reversedns_lookup.json` should also support `error` field used in Python part
5. We usually don't need to add additional records to CHANGELOG and update version number in the scope of one PR as action hasn't been merged yet. You can leave it at `0.1.0 - inital release` for now.

Thanks!

**Comment by KhadimBamba27** [comment]
Hello, 

Thank you for your valuable feedback! I have refactored the module according to your guidelines:
- Moved the module to `automation_modules/DNS`.
- Renamed the action to `action_dns_reverse_search` and the class to `DnsReverseSearchAction` for consistency and future DNS extensions.
- Removed the duplicate configuration file.
- Updated `action_dns_reverse_search.json` to include the `error` field in the results schema.
- Reset the version and CHANGELOG to `0.1.0` (Initial release).

The PR is now updated and ready for another review. Thanks!

**Review by lvoloshyn-sekoia — CHANGES_REQUESTED** [review]
Hey @KhadimBamba27 

Thanks for you fixes! Few more things:
1. Can you move `DNS` folder higher in the file tree? Right now it's `automation-library/automation_modules/DNS`, should be just `automation-library/DNS`
2. Could you add a logo (see [this](https://github.com/SEKOIA-IO/automation-library/blob/develop/Utils/logo.svg) generic logo) and a Dockerfile ([example](https://github.com/SEKOIA-IO/automation-library/blob/develop/Utils/Dockerfile))?
3. Could you also set python version to 3.11 in `pyproject.toml` in order to be the same with Dockerfile?

Thanks!

**Comment by KhadimBamba27** [comment]
Hello,

Thanks again for the review! I have applied the requested changes:
- Moved the `DNS` directory directly to the root of the repository (`automation-library/DNS`).
- Added the `logo.png` and a standardized `Dockerfile` matching the module's structure.
- Updated `pyproject.toml` to lock the Python version to `^3.11` to match the Docker base image.

Everything is pushed and ready for your final validation!

File: DNS/action_dns_reverse_search.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "Reverse DNS Lookup",
+  "description": "Resolve an IP address into its corresponding hostname using a PTR request",
+  "uuid": "1a2b3c4d-5678-90ab-cdef-1234567890ab",
+  "slug": "action_dns_reverse_search",
+  "docker_parameters": {
+    "command": "python main.py"
+  },
```
Comment (lvoloshyn-sekoia) [inline DNS/action_dns_reverse_search.json]: You just need to use the string you assigned in the `main.py`
```suggestion
  "docker_parameters": "action_dns_reverse_search",
```

File: DNS/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+FROM python:3.11-slim
```
Comment (lvoloshyn-sekoia) [inline DNS/Dockerfile]: Could you please just copy over [this standard Dockerfile](https://github.com/SEKOIA-IO/automation-library/blob/develop/Utils/Dockerfile) as is?

File: DNS/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+[tool.poetry]
+name = "dns"
+version = "0.1.0"
+description = "DNS automation module for Sekoia.io"
+authors = ["Khadim Bamba <khadimkd004@gmail.com>"]
+
+[tool.poetry.dependencies]
+python = "^3.11"
+sekoia-automation-sdk = "^1.12.0"
```
Comment (lvoloshyn-sekoia) [inline DNS/pyproject.toml]: Could you please upgrade `sekoia-automatin-sdk` to the latest version? You also need to include generated `poetry.lock` in the PR

**Review by lvoloshyn-sekoia — CHANGES_REQUESTED** [review]
Hey @KhadimBamba27 ,
thanks you for the fixes, I believe there are just a few more 😅

**Comment by KhadimBamba27** [comment]
Hello @lvoloshyn-sekoia,

Done! 
- I have copied your standard `Dockerfile` exactly as requested.
- Upgraded `sekoia-automation-sdk` to the latest version using Poetry.
- Generated and included the updated `poetry.lock` file in the repository.

Everything is pushed and ready for your review. Thanks!

**Review by lvoloshyn-sekoia — CHANGES_REQUESTED** [review]
Hey @KhadimBamba27 ,
with CI tests I found and fixed some non-obvious issues. However, we still have few issues remaining.

Could you please apply the `black` linter utility in order to format files in this new module? It's needed to pass our checks

**Comment by KhadimBamba27** [comment]
Hello @lvoloshyn-sekoia,

I have successfully run the `black` linter utility across the module files to format the Python code properly.

The updated and formatted files are now pushed. Thanks for your help with the CI fixes!

File: DNS/dns_modules/action_dns_reverse_search.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import socket
+from sekoia_automation.action import Action
+
```
Comment (lvoloshyn-sekoia) [inline DNS/dns_modules/action_dns_reverse_search.py:3]: Could you create empty `__init__.py` in the same directory?

File: DNS/action_dns_reverse_search.json (hunk 1 above)
Comment (lvoloshyn-sekoia) [inline DNS/action_dns_reverse_search.json]:
```suggestion
  "docker_parameters": "action_dns_reverse_search",
```

File: DNS/main.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+from sekoia_automation.module import Module
+from dns_modules.action_dns_reverse_search import DnsReverseSearchAction
+
+if __name__ == "__main__":
+    module = Module()
+    module.register_action("action_dns_reverse_search", DnsReverseSearchAction)
```
Comment (lvoloshyn-sekoia) [inline DNS/main.py]:
```suggestion
    module.register(DnsReverseSearchAction, "action_dns_reverse_search")
```

File: DNS/manifest.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "DNS",
+  "version": "0.1.0",
+  "description": "Module for DNS-related automations and lookups",
+  "uuid": "7a8b9c0d-1234-5678-90ab-cdef12345678",
+  "slug": "dns",
+  "actions": [
+    "action_dns_reverse_search"
+  ]
+}
```
Comment (lvoloshyn-sekoia) [inline DNS/manifest.json:16]:
```suggestion
{
  "configuration": {
    "type": "object",
    "title": "Module configuration",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "properties": {}
  },
  "description": "Module for DNS-related automations and lookups",
  "name": "DNS",
  "uuid": "7a8b9c0d-1234-5678-90ab-cdef12345678",
  "slug": "dns",
  "version": "0.1.0",
  "categories": [
    "Network"
  ]
}

```

File: DNS/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,16 @@
+FROM python:3.11
+
+WORKDIR /app
+
+RUN pip install poetry
+
+# Install dependencies
+COPY poetry.lock pyproject.toml /app/
+RUN poetry config virtualenvs.create false && poetry install --only main
```
Comment (lvoloshyn-sekoia) [inline DNS/Dockerfile]:
```suggestion
RUN poetry config virtualenvs.create false && poetry install --only main --no-root
```

File: DNS/pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+[tool.poetry]
+name = "dns"
+version = "0.1.0"
+description = "DNS automation module for Sekoia.io"
+authors = ["Khadim Bamba <khadimkd004@gmail.com>"]
+
```
Comment (lvoloshyn-sekoia) [inline DNS/pyproject.toml]:
```suggestion
package-mode = false
```

**Review by lvoloshyn-sekoia — CHANGES_REQUESTED** [review]
Hey @KhadimBamba27 ,
unfortunately you override my changes and they are no longer present. I've added them as comments in order to reproduce

File: DNS/dns_modules/action_dns_reverse_search.py (hunk 4 above)
Comment (KhadimBamba27) [reply DNS/dns_modules/action_dns_reverse_search.py:3]: I have successfully accepted all your suggestions directly on GitHub and synchronized my local environment. 

I also created the missing empty `__init__.py` file inside the `dns_modules` directory and pushed it. The module structure and formatting are now completely clean and ready for final review. Thank you so much for your guidance throughout this process!

File: DNS/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+description = "DNS automation module for Sekoia.io"
+authors = ["Khadim Bamba <khadimkd004@gmail.com>"]
+package-mode = false
+[tool.poetry.dependencies]
+python = "^3.11"
+sekoia-automation-sdk = "^1.23.0"
+
+[build-system]
+requires = ["poetry-core>=1.0.0"]
+build-backend = "poetry.core.masonry.api"
```
Comment (lvoloshyn-sekoia) [inline DNS/pyproject.toml:13]:
```suggestion
build-backend = "poetry.core.masonry.api"

[tool.poetry.dev-dependencies]
pytest = "*"
pytest-cov = "*"

```

**Review by lvoloshyn-sekoia — CHANGES_REQUESTED** [review]
Could you add suggested content to pyproject.toml and regenerate lock file?

File: DNS/pyproject.toml (hunk 9 above)
Comment (KhadimBamba27) [reply DNS/pyproject.toml:13]: The updated lock file is now pushed and ready. Thank you!

**Review by lvoloshyn-sekoia — CHANGES_REQUESTED** [review]
Seems finally CI working as it should. But now we need a test to pass 5% coverage

Could you please add a simple test for the action, while mocking network interaction? Here is [an example](https://github.com/SEKOIA-IO/automation-library/blob/develop/Utils/tests/test_action_utils_wait.py)

**Comment by KhadimBamba27** [comment]
Hello @lvoloshyn-sekoia,

I have added the unit tests for the `DnsReverseSearchAction` inside a new `tests/` directory. 

Following your example, I used `unittest.mock.patch` to mock the `socket.gethostbyaddr` interactions for both successful resolutions and `socket.herror` failure scenarios. I also ran the `black` linter on the test files.

The code coverage requirements should now be fully satisfied. Thanks!

**Review by lvoloshyn-sekoia — APPROVED** [review]

**Comment by KhadimBamba27** [comment]
Hello @lvoloshyn-sekoia,

Thank you for approving the changes! 

To satisfy the Codecov target threshold, I have expanded the test suite to cover all remaining edge cases: the missing `ip_address` argument validation and the catch-all `Exception` block.

With these additional test cases, the coverage for the new code should now reach 100% and clear the check. Thanks!

**Comment by KhadimBamba27** [comment]
Hello @lvoloshyn-sekoia,

Great news! All 132 tests and Codecov checks have successfully passed, and the branch is now fully compliant with green status.

The PR is ready to be merged whenever you are available. Thank you again for your time, review, and valuable guidance!
